nucleus/async_utils.py

```diff
@@ -0,0 +1,211 @@
+                filename=field.filename,
+                value=field.value,
+                content_type=field.content_type,
+            )
+        return form
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        for file_pointer in self._file_pointers:
+            file_pointer.close()
+
+
+def get_event_loop():
+    try:
+        loop = asyncio.get_event_loop()
+    except RuntimeError:  # no event loop running:
+        loop = asyncio.new_event_loop()
+    else:
+        nest_asyncio.apply(loop)
+    return loop
+
+
+def make_many_form_data_requests_concurrently(
+    client: "NucleusClient",
+    requests: Sequence[FormDataContextHandler],
+    route: str,
+    progressbar: tqdm,
+    concurrency: int = 30,
+):
+    """
+    Makes an async post request with form data to a Nucleus endpoint.
+
+    Args:
+        client: The client to use for the request.
+        requests: Each requst should be a FormDataContextHandler object which will
+            handle generating form data, and opening/closing files for each request.
+        route: route for the request.
+    """
+    loop = get_event_loop()
+    return loop.run_until_complete(
+        form_data_request_helper(
+            client, requests, route, progressbar, concurrency
+        )
+    )
+
+
+async def form_data_request_helper(
+    client: "NucleusClient",
+    requests: Sequence[FormDataContextHandler],
+    route: str,
+    progressbar: tqdm,
+    concurrency: int = 30,
+):
+    """
+    Makes an async post request with files to a Nucleus endpoint.
+
+    Args:
+        client: The client to use for the request.
+        requests: Each requst should be a FormDataContextHandler object which will
+            handle generating form data, and opening/closing files for each request.
+        route: route for the request.
+    """
+    async with aiohttp.ClientSession() as session:
+        tasks = [
+            asyncio.ensure_future(
+                _post_form_data(
+                    client=client,
+                    request=request,
+                    route=route,
+                    session=session,
+                    progressbar=progressbar,
+                )
+            )
+            for request in requests
+        ]
+        return await gather_with_concurrency(concurrency, *tasks)
+
+
+async def _post_form_data(
+    client: "NucleusClient",
+    request: FormDataContextHandler,
+    route: str,
+    session: aiohttp.ClientSession,
+    progressbar: tqdm,
+):
+    """
+    Makes an async post request with files to a Nucleus endpoint.
+
+    Args:
+        client: The client to use for the request.
+        request: The request to make (See FormDataContextHandler for more details.)
+        route: route for the request.
+        session: The session to use for the request.
+    """
+    endpoint = f"{client.endpoint}/{route}"
+
+    logger.info("Posting to %s", endpoint)
+
+    for sleep_time in RetryStrategy.sleep_times + [-1]:
```

Comment:
We should probably add jitter to these as well. Otherwise you could get a cascade of retries at the same instant. 

At some point @phil-scale suggested using https://github.com/jd/tenacity. There might be a reason to not implement details  of proper retrying 🙂

Reply:
I looked at tenacity and I didn't see a way of retrying on only specific statuses.